runtime/src/utils/buffer/write.rs

```diff
@@ -2,104 +2,116 @@ use crate::{Blob, Error, RwLock};
-    async fn flush(&mut self) -> Result<(), Error> {
-        if self.buffer.is_empty() {
-            return Ok(());
+    /// Panics if the end offset of the requested data falls outside the range of the logical blob.
+    fn read(&self, buf: &mut [u8], offset: u64) -> usize {
+        assert!(offset + buf.len() as u64 <= self.size());
+        if offset + buf.len() as u64 <= self.offset {
+            // Range does not overlap with the buffer.
+            return buf.len();
         }
 
-        // Take the buffer contents and write to blob
-        let buf = std::mem::take(&mut self.buffer);
-        let len = buf.len() as u64;
-        self.blob.write_at(buf, self.position).await?;
+        let (start, remaining) = if offset < self.offset {
+            // Some data is before the buffer.
+            (0, (self.offset - offset) as usize)
+        } else {
+            // Can read entirely from the buffer.
+            ((offset - self.offset) as usize, 0)
+        };
 
-        // Advance position to the new tip and reset buffer
-        self.position += len;
-        self.buffer = Vec::with_capacity(self.capacity);
-        Ok(())
-    }
+        let end = start + buf.len() - remaining;
+        assert!(end <= self.data.len());
 
-    /// Flushes buffered data and ensures it is durably persisted to the underlying [Blob].
-    ///
-    /// Returns an error if either the flush or sync operation fails.
-    async fn sync(&mut self) -> Result<(), Error> {
-        self.flush().await?;
-        self.blob.sync().await
+        // Copy the requested buffered data into the appropriate part of the user-provided slice.
+        buf[remaining..].copy_from_slice(&self.data[start..end]);
+
+        remaining
     }
 
-    /// Closes the writer and ensures all buffered data is durably persisted to the underlying [Blob].
-    ///
-    /// Returns an error if the close operation fails.
-    async fn close(&mut self) -> Result<(), Error> {
-        // Ensure all buffered data is durably persisted
-        self.sync().await?;
+    /// Merges the provided `data` into the buffer at the provided blob `offset` if it falls
+    /// entirely within the range `[buffer.offset,buffer.offset+capacity)`. The buffer will be
+    /// expanded if necessary to accommodate the new data, and any gaps that may result are filled
+    /// with zeros. Returns `true` if the merge was performed, otherwise the caller is responsible
+    /// for continuing to manage the data.
+    fn merge(&mut self, data: &[u8], offset: u64) -> bool {
+        let end_offset = offset + data.len() as u64;
+        let can_merge_into_buffer =
+            offset >= self.offset && end_offset <= self.offset + self.capacity as u64;
+        if !can_merge_into_buffer {
+            return false;
+        }
+        let start = (offset - self.offset) as usize;
+        let end = start + data.len();
+        // Expand buffer if necessary (fills with zeros).
```

Comment:
nit: below you `\n` before comment, may make sense to do the same here (or remove the below \n)

Reply:
yep added \n.